module-2/backend/week_four.md

```diff
@@ -10,14 +10,38 @@ Note: When you're done, submit a PR with a reflection in the comments about how
 ### Questions
 
 * What is a cookie?
+ A way to maintain state over HTTP request/response cycle.
+
 * What’s the difference between a session and a cookie?
+ A session is similiar to  a cookie except it allows you to have authorization/authentication.
+
 * What’s a flash and when do you want to use flashes?
+ Flash is a piece of functionality you get with rails. It lets you display temporary messages to your users.
+
 * Why do people say “HTTP is stateless”?
+ As a result of being  a RESTful app. It doesn't hold information during communication transactions and after they end.
+
 * What’s authentication? Explain.
+ Verifies a user is a specific user. Allows for relative user privacy across your app and for them to access the info specific to them.
+
 * What’s the difference between authentication and authorization?
+  The former confirms who they say they are, the latter deals with what permissions and access they have once authenticated.
+
 * What’s a before filter?
+  It controls what controller functionality your users/admin have access to you depending on what you specify.
+
 * How do we keep track of a user once they’ve logged in?
+  By assigning them to a session.
+
 * When do you want to namespace a resource? When do you want to nest a resource? What's the differences between those two approaches?
+  Nesting deals with an ownership attribute between two related resources. It changes your routes/URI patterns. Namespacing deals with authorization and you controlling who sees what. It does similiar actions to nesting and additionally gives you specific controller actions (per your namespace).
+
 * At a high level, what tools can you use to implement authorization? How would you use them?
+  Through assignment of specific roles to your users. Certain users will have higher privelages than others depending on the role assigned to them. An enforcement mechanism could be through the use of an enum.
```

Comment:
Main tools that we've used up to this point are `before_action`, storing user ids in our session, and inheritance to allow us to apply those `before_action`s to multiple controllers. We're using enums to get some friendly methods: `?` and `!`, as well as to save some space in the DB, but that's a nice-to-have, not the primary driver.